python/ray/experimental/workflow/workflow_storage.py

```diff
@@ -0,0 +1,229 @@
+
+    def inspect_step(self, step_id: StepID) -> StepInspectResult:
+        """
+        Get the status of a workflow step. The status indicates whether
+        the workflow step can be recovered etc.
+
+        Args:
+            step_id: The ID of a workflow step
+
+        Returns:
+            The status of the step.
+        """
+        field_list = self._storage.get_step_status(self._workflow_id, step_id)
+        # does this step contains output checkpoint file?
+        if field_list.output_object_exists:
+            return StepInspectResult(output_object_valid=True)
+        # do we know where the output comes from?
+        if field_list.output_metadata_exists:
+            return StepInspectResult(
+                output_step_id=self._locate_output_step_id(step_id))
+
+        # read inputs metadata
+        try:
+            metadata = self._storage.load_step_input_metadata(
+                self._workflow_id, step_id)
+            input_object_refs = metadata["object_refs"]
+            input_workflows = metadata["workflows"]
+        except (FileNotFoundError, json.JSONDecodeError):
+            input_object_refs = None
+            input_workflows = None
+        return StepInspectResult(
+            args_valid=field_list.args_exists,
+            func_body_valid=field_list.func_body_exists,
+            object_refs=input_object_refs,
+            workflows=input_workflows,
+        )
+
+    def _write_step_inputs(self, step_id: StepID, inputs: WorkflowInputs):
+        """Save workflow inputs."""
+        f = inputs.func_body
+        metadata = {
+            "name": f.__module__ + "." + f.__qualname__,
+            "object_refs": inputs.object_refs,
+            "workflows": inputs.workflows,
+        }
+        with serialization_context.workflow_args_keeping_context():
+            # TODO(suquark): in the future we should write to storage directly
+            # with plasma store object in memory.
+            args_obj = ray.get(inputs.args)
+        self._storage.dump_step_input_metadata(self._workflow_id, step_id,
+                                               metadata)
+        self._storage.dump_step_func_body(self._workflow_id, step_id, f)
+        self._storage.dump_step_args(self._workflow_id, step_id, args_obj)
+
+    def commit_step(self, step_id: StepID, ret: Union[Workflow, Any],
+                    forward_output_to: Optional[StepID]) -> None:
+        """When a workflow step returns,
+        1. If the returned object is a workflow, this means we are a nested
+           workflow. We save the DAG of the nested workflow.
+        2. Otherwise, checkpoint the
+
+        Args:
+            step_id: The ID of the workflow step. If it is an empty string,
+                it means we are in the workflow job driver process.
+            ret: The returned object from a workflow step.
+            forward_output_to: The output should also forward to the step
```

Comment:
I'm finding the terminology of forwarding quite confusing. I don't have a concrete suggestion here, but could we find an easier to understand term, or otherwise add some examples here to clarify what this means (what's the difference between step_id, forward_output_to, and shortcut_output_id?)